category/execution/ethereum/evm_test.cpp

```diff
@@ -476,107 +511,117 @@ TEST(Evm, deploy_contract_code)
-            auto const r2 = deploy_contract_code<EvmTraits<EVMC_HOMESTEAD>>(
-                s, a, std::move(r));
+        else {
+            // Fail to deploy code - out of gas (EIP-2)
             EXPECT_EQ(r2.status_code, EVMC_OUT_OF_GAS);
-            EXPECT_EQ(r2.gas_left, 700);
             EXPECT_EQ(r2.create_address, 0x00_address);
         }
     }
+}
 
-    // Spurious Dragon
-    {
-        uint8_t const ptr[25000]{0x00};
-        byte_string code{ptr, 25000};
+TYPED_TEST(MonadEvmRevisionTest, deploy_contract_code_max_code_size)
+{
+    static constexpr auto a{0xbebebebebebebebebebebebebebebebebebebebe_address};
 
-        State s{bs, Incarnation{0, 4}};
+    InMemoryMachine machine;
+    mpt::Db db{machine};
+    db_t tdb{db};
+    vm::VM vm;
+    commit_sequential(
+        tdb,
+        StateDeltas{{a, StateDelta{.account = {std::nullopt, Account{}}}}},
+        Code{},
+        BlockHeader{});
+    BlockState bs{tdb, vm};
+
+    if constexpr (
+        TestFixture::Trait::max_code_size() <
+        std::numeric_limits<size_t>::max()) {
+        uint8_t const ptr[250000]{0x00};
+        byte_string code{ptr, 250000};
+        static_assert(TestFixture::Trait::max_code_size() < 250000);
+
+        State s{bs, Incarnation{0, 0}};
 
         evmc::Result r{
             EVMC_SUCCESS,
             std::numeric_limits<int64_t>::max(),
             0,
             code.data(),
             code.size()};
-        auto const r2 = deploy_contract_code<EvmTraits<EVMC_SPURIOUS_DRAGON>>(
+        auto const r2 = deploy_contract_code<typename TestFixture::Trait>(
             s, a, std::move(r));
         EXPECT_EQ(r2.status_code, EVMC_OUT_OF_GAS);
         EXPECT_EQ(r2.gas_left, 0);
         EXPECT_EQ(r2.create_address, 0x00_address);
     }
+}
 
-    // London
-    {
+TYPED_TEST(MonadEvmRevisionTest, deploy_contract_code_validation)
+{
+    static constexpr auto a{0xbebebebebebebebebebebebebebebebebebebebe_address};
+
+    InMemoryMachine machine;
+    mpt::Db db{machine};
+    db_t tdb{db};
+    vm::VM vm;
+    commit_sequential(
+        tdb,
+        StateDeltas{{a, StateDelta{.account = {std::nullopt, Account{}}}}},
+        Code{},
+        BlockHeader{});
+    BlockState bs{tdb, vm};
+
+    // EIP-3541 validation
+    if constexpr (TestFixture::Trait::evm_rev() >= EVMC_LONDON) {
         byte_string const illegal_code{0xef, 0x60};
 
-        State s{bs, Incarnation{0, 5}};
+        State s{bs, Incarnation{0, 0}};
 
         evmc::Result r{
             EVMC_SUCCESS, 1'000, 0, illegal_code.data(), illegal_code.size()};
-        auto const r2 =
-            deploy_contract_code<EvmTraits<EVMC_LONDON>>(s, a, std::move(r));
+        auto const r2 = deploy_contract_code<typename TestFixture::Trait>(
+            s, a, std::move(r));
         EXPECT_EQ(r2.status_code, EVMC_CONTRACT_VALIDATION_FAILURE);
         EXPECT_EQ(r2.gas_left, 0);
         EXPECT_EQ(r2.create_address, 0x00_address);
     }
```

Comment:
Worth adding the pre-London case here that we _can_ deploy a contract starting with `0xEF`?